### PR TITLE
chore: send workspace pubsub events by owner id

### PR DESCRIPTION
**Pull request by ethanndickson**
We currently send empty payloads to pubsub channels of the form `workspace:<workspace_id>` to notify listeners of updates to workspaces (such as for refreshing the workspace dashboard).

To support https://github.com/coder/coder/issues/14716, we'll instead send `WorkspaceEvent` payloads to pubsub channels of the form `workspace_owner:<owner_id>`. This enables a listener to receive events for all workspaces owned by a user.
This PR replaces the usage of the old channels without modifying any existing behaviors.

```
type WorkspaceEvent struct {
	Kind        WorkspaceEventKind `json:"kind"`
	WorkspaceID uuid.UUID          `json:"workspace_id" format:"uuid"`
	// AgentID is only set for WorkspaceEventKindAgent* events
	// (excluding AgentTimeout)
	AgentID *uuid.UUID `json:"agent_id,omitempty" format:"uuid"`
}
```

We've defined `WorkspaceEventKind`s based on how the old channel was used, but it's not yet necessary to inspect the types of any of the events, as the existing listeners are designed to fire off any of them.

```
WorkspaceEventKindStateChange     WorkspaceEventKind = "state_change"
WorkspaceEventKindStatsUpdate     WorkspaceEventKind = "stats_update"
WorkspaceEventKindMetadataUpdate  WorkspaceEventKind = "mtd_update"
WorkspaceEventKindAppHealthUpdate WorkspaceEventKind = "app_health"

WorkspaceEventKindAgentLifecycleUpdate  WorkspaceEventKind = "agt_lifecycle_update"
WorkspaceEventKindAgentLogsUpdate       WorkspaceEventKind = "agt_logs_update"
WorkspaceEventKindAgentConnectionUpdate WorkspaceEventKind = "agt_connection_update"
WorkspaceEventKindAgentLogsOverflow     WorkspaceEventKind = "agt_logs_overflow"
WorkspaceEventKindAgentTimeout          WorkspaceEventKind = "agt_timeout"
```

### CODE REVIEW

**Comment by ethanndickson** [comment]
* **#14847** <a href="https://app.graphite.dev/github/pr/coder/coder/14847?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14792** <a href="https://app.graphite.dev/github/pr/coder/coder/14792?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14964** <a href="https://app.graphite.dev/github/pr/coder/coder/14964?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ethanndickson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/agentapi/api.go [hunk 1]
```diff
@@ -75,10 +78,6 @@ type Options struct {
 	ExternalAuthConfigs       []*externalauth.Config
 	Experiments               codersdk.Experiments
 
-	// Optional:
-	// WorkspaceID avoids a future lookup to find the workspace ID by setting
-	// the cache in advance.
-	WorkspaceID          uuid.UUID
```
Comment (ethanndickson) [inline coderd/agentapi/api.go:81]: This doesn't need to be optional anymore, presumably since we removed the v1 agent api.

File: coderd/wspubsub/wspubsub.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+type WorkspaceEvent struct {
+	Kind        WorkspaceEventKind `json:"kind"`
+	WorkspaceID uuid.UUID          `json:"workspace_id" format:"uuid"`
+
+	// WorkspaceName is only set for WorkspaceEventTypeStateChange
+	WorkspaceName *string `json:"workspace_name"`
+	// Transition is only set for WorkspaceEventTypeStateChange
+	Transition *database.WorkspaceTransition `json:"transition,omitempty"`
+	// JobStatus is only set for WorkspaceEventTypeStateChange
+	JobStatus *database.ProvisionerJobStatus `json:"job_status,omitempty"`
```
Comment (spikecurtis) [inline coderd/wspubsub/wspubsub.go:39]: This smells to me like you're going to use these values to determine the state updates that we send over the API. That's a bad idea because PubSub updates are not strictly ordered with respect to database transactions.  So you could have an update setting the JobStatus to Running, then an update setting it to Succeeded, but the PubSub messages could come Succeeded, Running.

The PubSub should not be relied on to give you the latest state, just the kick to go and query the database for the state.

File: coderd/agentapi/api.go [hunk 3]
```diff
@@ -242,6 +232,11 @@ func (a *API) publishWorkspaceUpdate(ctx context.Context, agent *database.Worksp
 		return err
 	}
 
-	a.opts.PublishWorkspaceUpdateFn(ctx, workspaceID)
+	a.opts.PublishWorkspaceUpdateFn(ctx, a.opts.OwnerID, wspubsub.WorkspaceEvent{
```
[the remark this answers is not in the document]
Comment (ethanndickson) [reply coderd/agentapi/api.go:201]: Broke it down further into:

```
WorkspaceEventKindStateChange     WorkspaceEventKind = "state_change"
WorkspaceEventKindStatsUpdate     WorkspaceEventKind = "stats_update"
WorkspaceEventKindMetadataUpdate  WorkspaceEventKind = "mtd_update"
WorkspaceEventKindAppHealthUpdate WorkspaceEventKind = "app_health"

WorkspaceEventKindAgentLifecycleUpdate  WorkspaceEventKind = "agt_lifecycle_update"
WorkspaceEventKindAgentLogsUpdate       WorkspaceEventKind = "agt_logs_update"
WorkspaceEventKindAgentConnectionUpdate WorkspaceEventKind = "agt_connection_update"
WorkspaceEventKindAgentLogsOverflow     WorkspaceEventKind = "agt_logs_overflow"
WorkspaceEventKindAgentTimeout          WorkspaceEventKind = "agt_timeout"
```

File: coderd/workspaces.go [hunk 4]
```diff
@@ -1708,7 +1724,9 @@ func (api *API) watchWorkspace(rw http.ResponseWriter, r *http.Request) {
 	defer cancelWorkspaceSubscribe()
 
 	// This is required to show whether the workspace is up-to-date.
-	cancelTemplateSubscribe, err := api.Pubsub.Subscribe(watchTemplateChannel(workspace.TemplateID), sendUpdate)
+	cancelTemplateSubscribe, err := api.Pubsub.Subscribe(watchTemplateChannel(workspace.TemplateID), func(ctx context.Context, msg []byte) {
+		sendUpdate(ctx, nil)
+	})
```
Comment (spikecurtis) [inline coderd/workspaces.go:1729]: why is this change necessary?

File: coderd/workspaceagents.go [hunk 5]
```diff
@@ -241,25 +242,20 @@ func (api *API) patchWorkspaceAgentLogs(rw http.ResponseWriter, r *http.Request)
...
-			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-				Message: "Internal error fetching workspace build job.",
-				Detail:  err.Error(),
-			})
-			return
-		}
-
-		api.publishWorkspaceUpdate(ctx, build.WorkspaceID)
+		api.publishWorkspaceUpdate(ctx, workspace.Workspace.OwnerID, wspubsub.WorkspaceEvent{
+			Kind:        wspubsub.WorkspaceEventKindAgentLogsUpdate,
```
Comment (spikecurtis) [inline coderd/workspaceagents.go:255]:
```suggestion
			Kind:        wspubsub.WorkspaceEventKindAgentLogsOverflow,
```

**Review by spikecurtis — APPROVED** [review]
